rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -19,237 +19,407 @@ The proposed layers should support ragged tensors.
-NUMERICAL_COLUMNS = ['age', 'fare']
-# input list to create functional model.
-model_inputs = []
-# input list to feed linear model.
-linear_inputs = []
-for feature_name in CATEGORICAL_COLUMNS:
-  feature_input = tf.keras.Input(shape=(1,), dtype=tf.string, name=feature_name, sparse=True)
-  vocab_list = sorted(dftrain[feature_name].unique())
-  # Map string values to indices
-  x = tf.keras.layers.IndexLookup(vocabulary=vocab_list, name=feature_name)(feature_input)
-  x = tf.keras.layers.CategoryEncoding(num_categories=len(vocab_list))(x)
-  linear_inputs.append(x)
-  model_inputs.append(feature_input)
-
-for feature_name in NUMERICAL_COLUMNS:
-  feature_input = tf.keras.Input(shape=(1,), name=feature_name)
-  linear_inputs.append(feature_input)
-  model_inputs.append(feature_input)
-
-linear_model = tf.keras.experimental.LinearModel(units=1)
-linear_logits = linear_model(linear_inputs)
-model = tf.keras.Model(model_inputs, linear_logits)
-
-model.compile('sgd', loss=tf.keras.losses.BinaryCrossEntropy(from_logits=True), metrics=['accuracy'])
-
-dataset = tf.data.Dataset.from_tensor_slices((
-  (tf.sparse.from_dense(dftrain.sex, "Unknown"), tf.sparse.from_dense(dftrain.n_siblings_spouses, -1),
-  tf.sparse.from_dense(dftrain.parch, -1), tf.sparse.from_dense(dftrain['class'], "Unknown"), tf.sparse.from_dense(dftrain.deck, "Unknown"),
-  tf.expand_dims(dftrain.age, axis=1), tf.expand_dims(dftrain.fare, axis=1)),
-  y_train)).batch(bach_size).repeat(n_epochs)
-
-model.fit(dataset)
+tf.feature_column.categorical_column_with_hash_bucket(key, hash_bucket_size)
+```
+to
+```python
+keras_input = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+hashed_input = tf.keras.experimental.preprocessing.Hashing(num_bins=hash_bucket_size)(keras_input)
 ```
 
-### Workflow 2
+Note the hashed output from KPL will be different than the hashed output from feature column, given how seed is choosen. `Hashing` also supports customized `salt`.
 
-The second example gives an instruction on how to transition from categorical feature columns to the proposed layers. Note that one difference for vocab categorical column is that, instead of providing a pair of mutually exclusive `default_value` and `num_oov_buckets` where `default_value` represents the value to map input to given out-of-vocab value, and `num_oov_buckets` represents value range of [len(vocab), len(vocab)+num_oov_buckets) to map input to from a hashing function given out-of-vocab value. In practice, we believe out-of-vocab values should be mapped to the head, i.e., [0, num_oov_tokens), and in-vocab values should be mapped to [num_oov_tokens, num_oov_tokens+len(vocab)).
+2. `tf.feature_column.categorical_column_with_identity`
+This feature column is merely for having identical inputs and outputs except mapping out-of-range value into `default_value`, thus can easily be done at data cleaning stage,
+not be part of feature engineering, and hence dropped in this proposal.
```

Comment:
nit: for better readability, do you want to move this to the end , since this is not replaced and deemed not needed in feature engineering.

Reply:
Done.

Reply:
sg!